첫 번째 리뷰 요청 드립니다!

MuseopKim opened the pull request:
안녕하세요!
첫 번째 리뷰 요청을 드립니다 :)

<br />

## 내용
### 회원가입, 로그인 API 구현
OAuth 인증을 먼저 구현하게 되어서 회원가입, 로그인 API 개발을 진행 하였습니다.
대략적인 인증 절차를 다음과 같이 설정 하고 진행 하였습니다.

<br />

### 인증 절차 및 방법

1. 클라이언트가 Access Token을 전달 받으면 이것을 가지고 리소스 서버에 사용자 데이터를 요청 합니다.
2. 리소스 서버로부터 사용자 데이터를 전달 받습니다. (OAuth ID, Name, Email, Avatar url, ...)
3. 가져온 사용자 데이터를 서버에 전달 합니다. 이때, 짧은 초단위 유효시간의 확인 토큰을 Authorization 헤더에 담아서 함께 전달 합니다.
4. 서버는 유효한 토큰인지를 검증하고, 전달 받은 사용자 데이터를 통해 기존에 존재하는 사용자인지 신규 사용자인지, 기존 사용자라면 데이터 변경 여부가 있는지를 판단하고 후처리를 합니다. (여기서의 변경 여부는, 사용자가 구글 혹은 네이버에서 변경한 데이터가 있는지 없는지 여부를 의미 합니다.)
5. 민감 정보를 제외한 사용자 데이터(DB id, name, avatar url, ...) 를 바탕으로 JWT를 생성 합니다.
6. 쿠키를 통해 JWT를 클라이언트로 전달 합니다.
7. 이후 권한이 필요한 모든 API 통신에 대하여 JWT를 쿠키에 담아 서버에 요청을 합니다.

<br />

## 특별히 봐줬으면 하는 부분 (질문)
프론트 개발자 분과 쿠키(httpOnly, Secure Cookie)를 통해 인증 토큰을 주고 받기로 했었는데요 :)
토큰을 Authorization 헤더를 통해 주고 받는 방식과, 쿠키를 통해 주고 받는 방식 중 어떤 것을 더욱 선호하시는지 궁금합니다.

(지금은 헤더를 통한 방식으로 수정을 하였는데, 아마 이 내용은 다음 PR때 코드에 반영 될 듯 합니다!)

---
초기 개발 일정을 빡빡하게 잡게 되어서, 급하게 진행하여 한번의 리뷰치고는 커밋이 좀 많습니다 😅
다음 PR부터는 자주, 작게 리뷰 요청을 드리도록 하겠습니다.

앞으로 잘 부탁 드립니다 🙇🏻‍♂️

MuseopKim committed:
chore: 로컬 환경 DB 설정 값 적용

[#2]

MuseopKim committed:
feat: Account Entity 추가

Provider
- OAuth provider 구분을 위해 추가

Role
- 시큐리티 권한 정보 관리를 위해 추가

[#2]

MuseopKim committed:
feat: Account 컬럼 기본 값 생성 하도록 `@PrePersist` 적용

- 계정 최초 생성 시 커리어 레터 제출 횟수에 대한 기본 값 설정
- soft deletes 구현을 위해 추가한 deleted 값 설정

MuseopKim committed:
feat: `@JpaAuditing` 설정 적용

BaseTimeEntity
- 자동 생성 시간을 갖도록 createdDateTime 필드 적용

[#2]

MuseopKim committed:
test: `@DataJpaTest` 관련 설정 추가

BaseRepositoryTest,
application.yml

MuseopKim committed:
test: Account 생성시 기본 값 테스트 (`@PrePersist`)

- submitCount, deleted 기본 값 설정 테스트

AccountRepository,
AccountRepositoryTest

MuseopKim committed:
test: Account - `@JpaAuditing` 테스트 추가

Account 생성 시간에 대한 테스트 추가

[#2]

MuseopKim committed:
refactor: AccountRepositoryTest 테스트 객체 분리

테스트시 given 절에서 발생하는 많은 중복을 제거하기 위해 테스트 객체 분리

AccountFactory
- Account 관련 테스트 객체 분리

[#2]

MuseopKim committed:
test: Account 기본 값 설정 테스트 - `@DisplayName` 수정

가독성을 위해 수정

[#2]

MuseopKim committed:
chore: Configuration property 바인딩을 위해 configuration-processor 추가

[#2]

MuseopKim committed:
feat: JWT 프로퍼티 값 추가

불변 객체를 유지하기 위해 `@ConstructorBinding` 적용

JwtProperties,
application.yml

CareerZipApplication
- `@ConstructorBinding` Enable

MuseopKim committed:
chore: StringUtils 의존성 추가

문자열을 보다 쉽게 다루기 위해 추가

[#2]

MuseopKim committed:
chore: JWT 토큰 라이브러리 추가

jjwt

MuseopKim committed:
fix: JwtProperties - issuer 필드 추가 및 타입 오류 수정

Issuer 필드 추가
expiration <-> prefix 타입 불일치로 인한 오류 수정

[#2]

MuseopKim committed:
feat: JwtTokenProvider 정의

JWT Token 전반에 대한 책임을 가지는 객체 정의
- issueToken() : 토큰 발행 메서드 정의
- validateAuthorization() : 최초 사용자 인증에 대한 Authorization 헤더 토큰을 검증하는 메서드 정의

[#2]

MuseopKim committed:
test: JwtTokenProvider 테스트 - 유효한 토큰 검증 테스트 추가

[#2]

MuseopKim committed:
test: Authorization Token 검증 실패 테스트 추가

JwtTokenProvider
- 테스트 과정에서 오류가 발견되어 수정

[#2]

MuseopKim committed:
test: SecretKey가 일치하지 않는 경우 실패하는 테스트 추가

[#2]

MuseopKim committed:
test: 만료된 JWT 토큰 검증 실패 테스트 추가

만료된 토큰으로 요청 할 경우 실패하는 테스트 추가
- expiredTokenTest()

[#2]

MuseopKim committed:
Refactor: JWT Token 관련 테스트 객체 분리

JwtFactory
- 추후 테스트 전역에서 발생 할 중복을 제거하기 위해 테스트 객체 생성 및 적용

[#2]

MuseopKim committed:
Feat: JWT 관련 예외 정의

jjwt가 기본적으로 제공하는 에외 객체를 그냥 쓰는 것이 아니라 구체적인 커스텀 예외 객체 던지도록 구현

JwtExpirationException,
InvalidJwtTokenException

[#2]

MuseopKim committed:
Test: 새로 정의한 JWT 관련 에러 객체 테스트 코드에 반영

MuseopKim committed:
Refactor: 토큰 검증 로직 가독성 개선

- 불가피한 try catch 구문에 의한 2depth 개선

MuseopKim committed:
Feat: Provider - 유효하지 않은 OAuth Provider를 검증하는 메서드 추가

메서드 및 관련 테스트 추가

[#2]

MuseopKim committed:
Feat: AccountRepository - Provider, OAuthId로 Account를 조회하는 메서드 및 관련 테스트 추가

사용자 인증 요청 시 등록된 OAuth 관련 데이터로(Provider, OAuthId) Account가 존재하는지 먼저 조회하기 위한 메서드

AccountRequest,
AccountRequestBuilder (Test)
- 인증한 사용자에 대한 데이터 DTO
- 프로덕션 코드에서 생성자를 막고, 테스트 코드에서 테스트 객체를 생성할 수 있도록 Builder 정의

Account
- AccountRequest DTO를 받아서 객체를 생성하도록 정적 팩토리 메서드 추가

EntityNotFoundException
- NotFoundException의 상위 예외 객체 정의
- TODO: 예외 세분화

[#2]

MuseopKim committed:
Fix: Account - 생성자에 접근제어자 private 처리

모든 Account 객체의 생성을 정적 팩토리 메서드를 통해서만 수행 하도록 의도를 나타내기 위해 변경

[#2]

MuseopKim committed:
Feat: AccountService - AccountRequest(DTO) 로부터 Account를 찾아 반환하는 메서드 추가

- 프론트에서 모든 OAuth 인증을 진행한 뒤 서버로 데이터를 반환하기 때문에 인증 요청시 사용자가 기존 DB에 등록 된 OAuth 사용자인지 아닌지 알수 없게 된다. 따라서 기존에 존재하는 사용자인지 아닌지 먼저 판별한 후, 기존 사용자가 아닐 경우 새로 생성하여 반환하도록 구현
- 관련 성공 케이스 테스트 적용

AuthException,
InvalidOAuthProviderException
- 인증 과정에서 올바른 OAuth Provider 정보로 데이터 등록 및 조회 요청을 하는 것이 아닐 경우 발생하는 예외 추가

[#2]

MuseopKim committed:
Test: AccountService.find() - 실패 테스트 케이스 추가

유효하지 않은 OAuth Provider를 포함하여 요청 할 경우 실패하는 테스트 케이스 적용

[#2]

MuseopKim committed:
Refactor: AccountRequest 테스트 객체 분리

given의 중복을 제거하기 위해 테스트 객체를 별도로 분리

MuseopKim committed:
Feat: API Response 객체 정의

프론트 개발자 분과 추후 다른 값을 고민 해보도록 한 뒤 우선 바디에 담겨질 본 data 값만 추가

[#2]

MuseopKim committed:
Fix: Account - ID 값이 필요한 테스트 환경에서 이를 생성하여 사용하기 위해 빌더에 id 필드 추가

- Account DTO 객체에 long이 primitive 타입이기 때문에 정적 팩토리 메서드가 관여 된 모든 테스트에서 Exception이 발생
- 테스트 이외의 환경에서는 정적 팩토리 메서드를 통해서만 생성 하도록 구현

[#2]

MuseopKim committed:
Fix: AccountService - 서비스 계층에서 엔티티가 아닌 DTO를 반환하도록 수정

연관된 테스트에 모두 반영

[#2]

MuseopKim committed:
Fix: AccountSummaryResponse -> AccountSummary DTO명 변경

이미 패키지가 DTO 하단에 있으므로 이름에서 Response 제외 가독성을 위해 변경

[#2]

MuseopKim committed:
Feat: AccountController - OAuth 인증 API 구현

클라이언트로부터 전달 받은 OAuth 사용자 데이터로부터 인증을 완료하는 API 구현

JwtProperties
- Long 타입 매핑을 지원하지 않으므로 String으로 바인딩 (이후 사용시 Long으로 Parsing)
- 쿠키를 사용하게 됨에 따라 쿠키 이름으로 변경

[#2]

MuseopKim committed:
Test: 인증 API 성공 테스트 추가

(+) 개발을 위해 모든 권한을 열어두도록 SecurityConfig 기본 권한 설정

MuseopKim committed:
Feat: Exception Handler 정의 (`@RestControllerAdvice`)

전역에서 예외를 잡도록 핸들러 정의

ErrorResponse
- API 에러 객체 정의

[#2]

MuseopKim committed:
Test: 인증 API - 유효하지 않은 JWT 토큰으로 요청했을 경우 실패하는 테스트 적용

테스트 Bad Request 케이스 적용

[#2]

MuseopKim committed:
Test: 인증 API - 만료된 JWT 토큰으로 요청 할 경우 실패하는 테스트 적용

[#2]

MuseopKim committed:
Chore: 패키지 재정리

각 도메인별로 따로 패키지화 하려던 것을 domain 하단으로 이동
(controller만 domain 밖으로 분리)

[#2]

MuseopKim committed:
Merge pull request #4 from Lubycon/feature/issue-2

인증 API

MuseopKim committed:
Chore: Slack notification 설정

[#1]

MuseopKim committed:
Merge pull request #6 from Lubycon/feature/issue-2

Chore: Slack notification 설정

MuseopKim committed:
Test: JPA 테스트를 제외한 테스트에서는 기본 값으로 ID를 제공하도록 테스트 객체 수정

RepositoryTest 이외의 테스트에서는 테스트 엔티티의 ID 값을 기본적으로 세팅 하도록 수정, 기존 테스트에 반영

[#2]

MuseopKim committed:
Chore: Slack notification checkout branch 정보 수정

[#2]

MuseopKim committed:
Chore: Slack notification checkout branch 수정

[#2]

evan-moon commented on .github/workflows/slack-notification.yml:
```diff
@@ -0,0 +1,16 @@
+      - name: Checkout
+        uses: actions/checkout@master
+      - name: Fire Notification
+        uses: Lubycon/github-reviewer-slack-noti-action@v1.2.0
```
```suggestion
        uses: Lubycon/github-reviewer-slack-noti-action@v1.2.1
```

무섭님 `1.2.0` 버전은 버그가 있어용! `1.2.1` 버전으로 올려주시면 감사하겠습니다! 🙏

MuseopKim replied:
처음이라 잘 몰랐는데 알려주셔서 감사합니다! 🙏

MuseopKim replied:
몰랐는데 바로 알려주셔서 감사합니다 동욱님! 🙏

MuseopKim committed:
Update .github/workflows/slack-notification.yml

버전 수정 (1.2.0 -> 1.2.1)

[#2]

Co-authored-by: Evan Moon <bboydart91@gmail.com>

MuseopKim committed:
Merge branch 'main' into dev

yeong-chae commented on src/main/java/com/careerzip/domain/account/dto/request/AccountRequest.java:
```diff
@@ -0,0 +1,24 @@
+@Getter
+public class AccountRequest {
+
+    private String provider;
```
DTO라면, final과 nullable 여부까지 어노테이션으로 걸어주시면 가독성이 더 좋을 것 같아유!
`@NotNull`, `@Nullable`

yeong-chae replied:
모든 DTO 공통입니다

yeong-chae replied:
Provider도 enum인게 좋을 것 같아용!

MuseopKim replied:
다음 수정 때 nullable을 DTO에 한번 적용 해보도록 하겠습니다!
그런데, Request DTO는 final을 적용할 수 없는 듯 한데, 응답 DTO에만 적용해도 충분한 것일까요?

yeong-chae replied:
objectMapper옵션설정에 따라서 가능할거에여!

MuseopKim replied:
영채님 힌트 주셔서 감사합니다! 적용 해보겠습니다ㅎㅎ

yeong-chae commented on src/main/java/com/careerzip/domain/account/dto/response/AccountSummary.java:
```diff
@@ -0,0 +1,34 @@
+    private final String name;
+    private final String email;
+    private final String avatarUrl;
+    private final String role;
```
role은 아마 타입이 한정되있을 것 같은데, enum으로 가는건 어떨까요?

MuseopKim replied:
DTO라서 아무 생각 없이 String으로 갔었는데, enum으로 수정 해보겠습니다!ㅎㅎ 
감사합니다.

shinsunyoung commented on src/main/java/com/careerzip/controller/AccountController.java:
```diff
@@ -0,0 +1,35 @@
+
+    @ResponseStatus(HttpStatus.OK)
+    @PostMapping("/auth")
+    public ApiResult<AccountSummary> authorize(@RequestHeader HttpHeaders headers, @RequestBody AccountRequest accountRequest,
```
리턴타입을 `ApiResult`로 커스텀하신 이유가 있으실까요? 가독성 때문인가욥??

MuseopKim replied:
프론트 개발자분께서 추후 200 응답에서 payload가 존재하지 않을 케이스가 있을 수 있으니, 이를 구분할 수 있는 어떤 방법이 있으면 좋겠다는 의견을 주셨었습니다ㅎㅎ

boolean 필드를 따로 둘까도 고민 했었는데, 우선 `data`라는 필드로 Null 체크를 통일된 방식으로 할 수 있도록 해두었던 것 같네요ㅎㅎ

추후 필요에 따라 다른 필드가 추가 될 수도 있는데, 우선 이렇게 정의 해두게 되었습니다ㅎㅎ

yeong-chae commented on src/main/java/com/careerzip/domain/account/entity/Account.java:
```diff
@@ -0,0 +1,81 @@
+package com.careerzip.domain.account.entity;
+
+import com.careerzip.domain.account.dto.request.AccountRequest;
```
Rquest용 DTO가 entity 레벨까지 내려오는건 결합도가 너무 높아보여요!
결합도를 낮춰보면 좋을 것 같아요~

yeong-chae commented on src/main/java/com/careerzip/domain/account/entity/Account.java:
```diff
@@ -0,0 +1,81 @@
+    private boolean deleted;
+
+    @Builder
+    private Account(Long id, String oauthId, Provider provider, String name, String email, String avatarUrl, Role role,
```
음 테스트용 코드가 프로덕션 코드에 껴있는건 좋지 않아보이는데, 해결방법이 딱히 떠오르질 않네요..

MuseopKim replied:
만약 이 값을 설정하지 않으면, Mocking을 매우 많이 해야 해서 비효율적이라 생각을 했었습니다 🥲
계속 고민 했었는데, 떠오르지 않아서 선택했던 방식이었어요 ㅎㅎ 완벽한 해결책은 아니었던 듯 해서 더 고민 해보겠습니다... 🙇🏻‍♂️

shinsunyoung commented on src/main/java/com/careerzip/domain/account/dto/request/AccountRequest.java:
```diff
@@ -0,0 +1,24 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
이 어노테이션을 다신 이유가 있으실까요??

MuseopKim replied:
아아 이건 테스트 코드 때문인데요ㅎㅎ 

요청 DTO는 Jackson2HttpMessageConverter 가 JSON을 객체로 변환 해주기 때문에 별도의 생성자나 setter를 정의하지 않아도 되지만, 테스트 코드 작성을 위해서는 요청 DTO에도 이를 생성할 수 있는 방법이 필요 합니다.

오직 테스트 때문에 불필요한 코드를 본 코드에 추가하는 것을 지양해야 한다는 생각을 가지고 있는데요 :)

이에 따라 프로덕션 코드에는 최대한 영향을 주지 않으면서 테스트 코드에서는 이를 생성할 수 있도록, 기본 생성자는 private으로 막아놓고, 테스트 코드의 같은 패키지에서는 정의 해둔 생성자를 호출 할 수 있도록 했습니다ㅎㅎ

이렇게 하면 패키지 외부에서는 생성자 호출이 막히기 때문에, 프로덕션 코드에는 영향을 최소화 하면서도 테스트를 위한 객체는 생성을 할 수 있게 됩니다!

shinsunyoung replied:
아아 답변 감사합니다!! 그런데 AccountRequest에 이미 생성자가 정의되어있던데, 그러면 어차피 기본 생성자 정의는 못하는거 아닌가욥??

MuseopKim replied:
선영님! 다른 생성자를 정의 했음에도 기본 생성자를 넣었던 이유는 `ObjectMapper`가 역직렬화를 수행할 때, 기본 생성자가 꼭 필요하다고 알고 있었기 때문이었는데요 :) 

간단히 찾아보니 `jackson-datatype-jdk8` 가 기본 생성자가 없으면 다른 생성자를 이용하도록 내부적으로 모듈을 등록 해준다고 하네요!
역직렬화 과정에서 문제가 되지 않는다면, 선영님 말씀처럼 기본 생성자를 굳이 정의 할 필요가 없을거 같습니다.

선영님 덕분에 잘 못 알고 있었던 것을 한번 짚어볼 수 있었습니다 😄 감사해요!

MuseopKim replied:
선영님! 다른 생성자를 정의 했음에도 기본 생성자를 넣었던 이유는 ObjectMapper가 역직렬화를 수행할 때, 기본 생성자가 꼭 필요하다고 알고 있었기 때문이었는데요 :)

간단히 찾아보니 jackson-datatype-jdk8 가 기본 생성자가 없으면 다른 생성자를 이용하도록 내부적으로 모듈을 등록 해준다고 하네요!
역직렬화 과정에서 문제가 되지 않는다면, 선영님 말씀처럼 기본 생성자를 굳이 정의 할 필요가 없을거 같습니다.

선영님 덕분에 잘 못 알고 있었던 것을 한번 짚어볼 수 있었습니다 😄 감사해요!

yeong-chae commented on src/main/java/com/careerzip/domain/account/entity/Account.java:
```diff
@@ -0,0 +1,81 @@
+                      .build();
+    }
+
+    @PrePersist
```
여기서 할필요 없이, 빌더나 프로퍼티 자체에 기본값을 선언해도 충분할 것 같슴니다

yeong-chae commented on src/main/java/com/careerzip/domain/account/service/AccountService.java:
```diff
@@ -0,0 +1,34 @@
+package com.careerzip.domain.account.service;
+
+import com.careerzip.domain.account.dto.request.AccountRequest;
+import com.careerzip.domain.account.dto.response.AccountSummary;
```
여기도 마찬가지에요. 프레젠테이션 레이어용 DTO가 어플리케이션 레이어까지 내려오는건 결합도가 높아지는 설계라서
해결하면 좋을 것 같슴니다~!

MuseopKim replied:
최대한 객체를 그대로 넘겨서 팩토리 메서드로 해결하고 싶었던 마음이었는데, 이 부분에 대해서는 계속 고민 해보겠습니다~! 👍

shinsunyoung commented on src/main/java/com/careerzip/domain/account/entity/Role.java:
```diff
@@ -0,0 +1,17 @@
+
+public enum Role {
+    MEMBER("ROLE_MEMBER"),
+    ADMIN("ROLE_ADMIN")
```
ADMIN은 왜 필요한건가욥?! 지금 코드를 봤을 때 어드민으로 분기하는 기능은 없는 것 같아서 질문 드립니당!

MuseopKim replied:
추후(가까운 미래에...?) ADMIN이라는 Role이 필요하게 될 것으로 예상 되어 미리 추가를 해두었었는데요! 선영님 말씀처럼 현재는 필요가 없는 코드는 맞습니다! 미리 정의 해 둔 값이에요 🙂

shinsunyoung replied:
음음 ADMIN이라는 값이 필요해지는 시점에 추가를 하는건 어떨까욥?! 나중에 ADMIN이 필요하지 않게 될 수도 있는데, 그런 경우에는 그대로 잊혀질 수도 있는 값이 될 수도 있을 것 같아서욥! 😂

MuseopKim replied:
선영님 말씀이 맞습니다! ㅎㅎ
제가 이 부분에 대해 다른 리뷰어 분께도 이전에 한번 지적을 받았던 적이 있는데, 요번 기회에 다시 한번 새기겠습니다 🙏
감사해요ㅎㅎㅎ

MuseopKim replied:
선영님 말씀이 맞습니다ㅎㅎ
제가 이 부분에 대해 다른 리뷰어 분께도 이전에 한번 지적을 받았던 적이 있는데, 이번 기회에 다시 한번 새기겠습니다 🙏 감사해요!

shinsunyoung commented on src/main/java/com/careerzip/domain/account/entity/Account.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @PrePersist
+    public void prePersist() {
```
요 메소드는 어디에 쓰는건가용?? 👀

MuseopKim replied:
`@PrePersist` 를 정의 해두면, 엔티티가 영속화 되기 바로 이전에 특정 로직들을 실행 하도록 할 수 있습니다.

저는 주로 기본 값을 설정할 때 (여기서는 summitCount, deleted 필드가 이에 해당이 되네요.)
 
`@ColumnDefinition` 으로 필드에 직접 default value를 정의 해도 되는데, 이 방법의 경우 DB가 변경될 경우 특정 vendor에 종속될 수 되는걸로 알고 있어서 이 방법을 더 선호하는거 같습니다ㅎㅎ

(그런데 위에 영채님께서, 기본 값은 빌더에서 처리해도 될 듯 하다는 피드백을 주셔서 변경을 고려 중이니 참고 해주세요!)

shinsunyoung replied:
아하 ㅎㅎ 저도 빌더에서 기본값을 적용해도 되지 않을까?! 라는 생각을 가지고 있어서 드렸던 질문이였어유 ㅎㅎ 감사합니다!

yeong-chae commented on src/main/java/com/careerzip/global/api/ApiResult.java:
```diff
@@ -0,0 +1,17 @@
+        this.data = data;
+    }
+
+    public static <T> ApiResult<T> SUCCESS(T data) {
```
요건 취향인 것 같긴 하지만, 스태틱 메서드라도 메서드가 대문자인건 좀 어색하네요..!

MuseopKim replied:
컨트롤러 전역에서 API마다 사용되는 표준 응답 객체라서, static import로 사용을 하기 때문에, 가독성을 위해 썼었는데요.
어색하시다는 느낌이 어떤 것인지 알거 같습니다! 변경을 고려 해보겠습니다... 😅

shinsunyoung commented on src/main/java/com/careerzip/domain/account/repository/AccountRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+
+    @Query("SELECT a FROM Account a WHERE a.provider = :provider AND a.oauthId = :oauthId")
```
네이티브 쿼리를 사용하셨는데, 최적화를 위해 네이티브 쿼리를 사용하신건가요?? 👀

MuseopKim replied:
Data JPA가 제공하는 쿼리 메서드를 사용해도 됐는데, 가독성을 해치는 경우가 많아서 정말 간단한 `ex. findById` 같은게 아니면 JPQL로 직접 정의하는 편인거 같습니다!

yeong-chae replied:
저도 종종 요런식으로 사용합니당ㅋㅋㅋ

shinsunyoung replied:
아하 저는 JPQL을 사용하면 type safe하지 않아서 JPQL 보다는 QueryDSL을 선호하는 편이에여!!
이건 개인 취향 차이인 것 같네여 ㅎㅎㅎ

MuseopKim replied:
아아! 맞습니다ㅎㅎ 
저도 이번 프로젝트에 QueryDSL을 적용 해볼 생각이에요! 😄

yeong-chae commented on src/main/java/com/careerzip/domain/account/service/AccountService.java:
```diff
@@ -0,0 +1,34 @@
+
+    private final AccountRepository accountRepository;
+
+    @Transactional
```
여기서 어떤 의미로 `@Transactional` 걸어두신거에여?

MuseopKim replied:
지금 현재 코드에는 반영이 되어 있지 않은데, 
(아주 깊은 부분까지 아는 것은 아니지만),`@Transactional(readOnly=true)`를 걸어두면 조회시에 하이버네트에서는 더티 체킹을 생략해서 (미미하지만) 최적화가 가능한 것으로 알고 있습니다 :)
그래서 개인적으로 서비스 계층에 클래스 레벨로 `@Transactional(readOnly=true)` 를 기본 설정으로 해놓고, DB에 변경점이 일어나는 메서드들에 대해서는 `@Transactional` 을 걸어두는 편입니다! 😅

(혹시 잘 못 알고 있는 부분이 있다면 조언을 부탁 드립니다 🙏)

yeong-chae replied:
굳굳! 혹시 관성으로 쓰시는게 아닐까 해서 여쭤봤어요ㅎㅎㅎ

yeong-chae commented on src/main/java/com/careerzip/global/error/response/ErrorResponse.java:
```diff
@@ -0,0 +1,19 @@
+        this.message = message;
+    }
+
+    public static ErrorResponse from(ErrorCode errorCode) {
```
요기는 소문자네여

MuseopKim replied:
`ApiResult`와 같이 에러를 담당하는 같은 표준 응답 객체인데, 여기서는 static import를 사용하지 않아서 가독성을 해치지 않는다고 생각하여 소문자로 사용 했던 것 같습니다! 🤔

shinsunyoung commented on src/main/java/com/careerzip/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,65 @@
+        return new Cookie(jwtProperties.getCookieName(), issueToken(account));
+    }
+
+    public boolean validateAuthorizationToken(String header) {
```
여기서 `header`가 `header`의 어떤 정보를 가져오는건가욥?? Request Header를 모두 String 형태로 가져오는건가욥??

MuseopKim replied:
`jwtTokenProvider.validateAuthorizationToken(headers.getFirst(HttpHeaders.AUTHORIZATION));`
 컨트롤러에서 해당 메서드를 실행하는 부분인데요!

HTTP 헤더 중, `Authorization`에 해당하는 값을 String으로 넘기도록 했습니다ㅎㅎ

yeong-chae replied:
파라미터명도 그럼 `authorizationHeader` 같은 좀 더 직관적인 파라미터명을 쓰면 좋을 것 같아요!

shinsunyoung replied:
저도 영채님 의견에 동의합니당!! 지금 `header`라는 변수명만 봤을 때에는 헤더 전체값을 모두 가져오는 것 같아 보일 수도 있을 것 같아용!!

MuseopKim replied:
요것도 다음 수정 때 반영 해볼게요! 감사합니다ㅎㅎ

shinsunyoung commented on src/main/java/com/careerzip/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,65 @@
+                .setSigningKey(jwtProperties.getSecretKey())
+                .parse(token);
+        } catch (ExpiredJwtException exception) {
+            throw new JwtExpirationException();
```
혹시 여기서 예외를 `JwtExpirationException`에게 위임한 이유가 있나욥?? `ExpiredJwtException`나 `JwtExpirationException`나 둘 다 같은 의미로 전달될 것 같아서욥!!

MuseopKim replied:
말씀하신 것처럼 커스텀 예외를 직접 던지기 위해서였는데요ㅎㅎ

try catch로 잡아야 하는 checked exception을 그냥 사용 할 경우, 상황을 명확히 드러내기 힘들다고 생각되어, 평소 이런 경우에는 구체적인 예외로 다시 던져주는 편입니다! 

상세한 예외 메시지를 정의하고, 클라이언트에 더 명확한 에러 정보를 주기 위해 구체적인 예외를 던지기 위한 의도였습니다! 😅

shinsunyoung replied:
아하 `ExpiredJwtException`가 checked exception이군여!!

shinsunyoung commented on src/main/java/com/careerzip/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,65 @@
+            throw new InvalidJwtTokenException();
+        }
+
+        return true;
```
`validateAuthorizationToken()`를 사용하는 곳을 찾아보았는데 true 값을 사용해서 처리하는 로직이 없는 것 같아요! (혹시 제가 못찾은거라면 말씀해주세용 😭) 혹시 return값을 boolean으로 두신 이유가 있으실까용??

MuseopKim replied:
이건 선영님 말씀처럼 반환 값이 필요가 없었던 메서드 같습니다!
알려주셔서 감사합니다! 👍👍👍

yeong-chae replied:
아주 꼼꼼쓰하신 선영님..,

shinsunyoung commented on src/main/java/com/careerzip/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private boolean isInvalidHeader(String header) {
+        return !header.startsWith("Bearer ");
```
마이너한 제안이긴 한데 `"Bearer"`라는 문자열을 상수로 가져가도 좋을 것 같아요!

MuseopKim replied:
좋은 제안 해주셔서 감사합니다! 다음 수정 때 반영 할게요ㅎㅎ

shinsunyoung commented on src/main/java/com/careerzip/security/SecurityConfig.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        // NOTICE: 개발의 편의성을 위해 csrf 옵션을 종료 합니다.
+        http.csrf().disable()
```
요건 나중에 변경할 코드인가요?!

MuseopKim replied:
아마 쿠키 사용을 하지 않을 예정인데, 이 옵션을 켜두면 시큐리티 때문에 `h2` 콘솔 접근에 문제가 생겨서 개발 단계에서는 꺼두는 편이에요! ㅎㅎ

shinsunyoung replied:
넵넵 프로덕트를 올릴 때에는 변경하시면 좋을 것 같네욥! 
IDE에 따라 다르지만 TODO, FIXME 주석헤더를 사용하시면 한 번에 모아볼 수 있으니 고려해보시는 것도 좋을 것 같아요!

shinsunyoung commented on src/test/java/com/careerzip/global/jwt/JwtTokenProviderTest.java:
```diff
@@ -0,0 +1,117 @@
+                .hasMessage(jwtExpiredError.getMessage());
+    }
+
+    @ParameterizedTest
```
요건 ParameterizedTest로 작성하신 이유가 있으실까요?!

MuseopKim replied:
현재 이 테스트에 대해서 2개의 Wrong 케이스 문자열을 정의 해둔 `@MethodSource` 를 사용하는데, 
이렇게 하면 각 문자열이 하나씩  테스트 메서드에 전달 되면서 2개의 값에 대해 한번에 테스트를 할 수 있게 됩니다!

만약 정말 다양한 경계 값 테스트 같은 것이 필요 할 경우, `@ParameterizedTest`와 `@MethodSource`를 사용하면 다양한 케이스 값에 대해 한번에 테스트 할 수 있는 장점이 있습니다ㅎㅎ

shinsunyoung reviewed (commented):
무섭님 코드 잘봤습니당!! 개인적으로 궁금한 점 몇가지 코멘트로 남겨두었어요! 👍👍👍

shinsunyoung commented on src/main/java/com/careerzip/config/jpa/JpaConfig.java:
```diff
@@ -0,0 +1,9 @@
+
+@EnableJpaAuditing
+@Configuration
+public class JpaConfig {
```
혹시 이 파일을 따로 분리하신 이유가 있으실까요?

MuseopKim replied:
이것도 테스트 코드를 위함인데요.

`@EnableJpaAuditing`을 적용하면 `@Entity` 가 적용된 클래스가 꼭 필요한걸로 알고 있습니다!
만약 이걸 Application에 직접 적용하면 컴포넌트 스캔 범위 때문에 
테스트 코드에서 에러가 발생하게 됩니다.

그래서 별도의 `@Configuration` 으로 분리 하게 되었습니다.

shinsunyoung replied:
테스트 코드일 때에는 컴포넌트 스캔 범위가 다른가요?? 😮

MuseopKim replied:
아아 Application에 직접 적용하면 `@WebMvcTest` 와 같이 `@Entity`와 전혀 관련이 없는 테스트에도 영향을 주게 되는 것으로 알고 있습니다. 그래서 꼭 분리를 해주어야 되더라구요! ㅎㅎ

shinsunyoung replied:
음음 그렇다면 Application에 `@EnableConfigurationProperties({JwtProperties.class})`를 붙이셨던데, 그러면 `@DataJpaTest` 같이 DB 테스트만 해야하는 부분은 JWT와 관련이 없는데 영향을 주는거 아닌가욥?? 제가 잘못이해한걸까요 ㅠㅠ ??

MuseopKim replied:
위 코멘트가 약간 선영님께 혼란을 드린 것 같은데요ㅎㅎ

Application에 `@EnableJpaAuditing`을 적용하면 `@Entity`가 존재하지 않는 `@WebMvcTest`의 컴포넌트 스캔 범위에도 영향을 주기 때문에 해당 애노테이션이 적용된 모든 테스트를 수행하지 못하고 에러가 발생하게 됩니다!ㅎㅎ

각 계층간에 논리적인 분리라기 보다는, 에러를 막기 위해 컴포넌트 스캔 범위를 분리 하는 것이라고 이해 해주시면 좋을거 같네요!

shinsunyoung replied:
아하 넵 이해했습니다 ㅎㅎ 감사해요 🙏

shinsunyoung commented on src/main/java/com/careerzip/domain/account/entity/Account.java:
```diff
@@ -0,0 +1,81 @@
+        this.deleted = deleted;
+    }
+
+    public static Account from(AccountRequest accountRequest) {
```
저는 이렇게 되면 엔티티가 DTO에 의존하는 코드가 생기게 되고, 레이어의 역행이 일어나는 상황이라고 생각하는데 혹시 다른 분들 생각은 어떠신지 궁금합니다...! (뇌피셜이에여)

MuseopKim replied:
이거 영채님께서도 같은 내용을 말씀 해주셔서 변경을 고민 하고 있어요ㅎㅎ 차라리 DTO에서 엔티티를 생성하도록 하게 할 생각인데, 다른 분들 코드를 보면서 (선영님 코드도?) 힌트를 얻어 볼 생각이에요 👍

좋은 지적 해주셔서 감사합니다!

shinsunyoung replied:
넵넵 저는 엔티티는 보통 빌더 패턴만 가져가는 편이에여!! ㅎㅎㅎ

shinsunyoung commented on src/main/java/com/careerzip/global/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    public boolean validateAuthorizationToken(String header) {
+        if (isInvalidHeader(header)) {
```
요것도 따로 메소드로 빼면 좋을 것 같아요! SRP의 관점에서 봤을 때 ... ?!

MuseopKim replied:
`isInvalidHeader(header)` 부분을 말씀 하시는게 맞을까요?!

shinsunyoung replied:
예외처리하는 부분이용!!

```
if (isInvalidHeader(header)) {
    throw new InvalidJwtTokenException();
}
```

MuseopKim replied:
아 이해 했습니다! 반영 해볼게요 👍

shinsunyoung commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,31 @@
+
+  jpa:
+    database-platform: org.hibernate.dialect.MySQL57InnoDBDialect
+    open-in-view: false
```
오 전 맨날 OSIV false로 설정하는거 맨날 까먹는데 안놓치시고 설정해두셨네욥 ... 👍👍 멋지심니다

shinsunyoung reviewed (commented):
코드 보다가 궁금한 점이 몇 가지 있어서 추가로 코멘트 달아놨슴니당 ㅎㅎㅎㅎ
추가로 무섭님 테스트 코드를 보니까 Mock 테스트를 많이 하시던데, 평소에 Mock 테스트를 선호하시는 편이신가욥??

MuseopKim commented:
@shinsunyoung 선영님, 코멘트 남겨주셔서 감사합니다ㅎㅎ 말씀 해주신 `레이어의 역행`에 대한 부분은 꼭 변경하는게 좋을 것 같네요 👍 좋은 지적 해주셔서 감사합니다!

그리고 질문 주신 내용에 대한 답변 드립니다.

현재 단위 테스트를 계속해서 적용하려 하고 있는데요ㅎㅎ
만약 통합 테스트가 아닌, 각 계층에 대한 단위 테스트의 경우, 의존성에 대한 Mocking이 불가피 하게 됩니다. 

통합 테스트가 아닌, 단위 테스트를 위해 Mock 객체를 적극 사용하게 되었다고 이해 해주시면 될 거 같습니다! 😄

shinsunyoung reviewed (commented):
무섭님 ㅋㅋㅋ 자꾸 코멘트를 추가로 달게 되네요 ... 😅  리뷰 내용을 이제 봐서 제 의견 공유 드립니다!

> 프론트 개발자 분과 쿠키(httpOnly, Secure Cookie)를 통해 인증 토큰을 주고 받기로 했었는데요 :)
토큰을 Authorization 헤더를 통해 주고 받는 방식과, 쿠키를 통해 주고 받는 방식 중 어떤 것을 더욱 선호하시는지 궁금합니다.

[공식 스펙](https://jwt.io/introduction)에는 헤더를 통해 주고 받는게 일반적인 스펙이라고 나와있어요! 이걸 물어보신게 맞으실까요?
```
Authorization: Bearer <token>
```
![image](https://user-images.githubusercontent.com/42836576/117582679-2c0bed80-b13e-11eb-9624-fe80d2586383.png)

MuseopKim commented:
@shinsunyoung 선영님! 공유 해주신 내용 읽어 보았습니다.
짧게 물어봐주셔서 선영님께서 질문 해주신 의도를 파악하지 못하고 피상적으로 답을 드린 것 같은데요 :)

이 부분에 대해 저는 취사 선택의 문제라고 생각 하는 것 같아요ㅎㅎ
(제 경우에는) 단위 테스트를 적용 할 수 있으면 최대한 적용 하되, 외부 API 라던지, 특정 환경에 종속 된다거나, 커버할 수 없다면 통합 테스트를 선택적으로 적용을 고려할 수 밖에 없을거 같다는 생각을 가지고 있는거 같습니다ㅎㅎ

공유 해주신 내용 처럼 Mock 객체가 말그대로 Mocking이기 때문에, 코드의 현실 상황을 항상 반영하지 못하고, 그렇기 때문에 잘못 사용하면 단점 역시 분명하다고 생각 합니다. 
아주 작은 예를 들면... 이런점들 때문에 이전에 음원 검색 API를 `@RestClientTest` 로 단위 테스트 적용 하려다가 스펙이 바뀌면 의미가 없는 테스트가 된다는 생각이 들어서, 이런 부분에는 통합 테스트를 적용하곤 했었네요!

Mock을 사용해서 얻을 수 있는 이점은 최대한 얻고, (의존성이라던지, 외부 환경이라던지) 서로간의 간섭이 일어나야 하는 부분에서는 과감히 통합 테스트를 적용 하는게 좋다는 생각이에요!

저는 이번 프로젝트에서 스프링 시큐리티를 계속 사용하게 될텐데, 아마 시큐리티 필터를 거쳐가게 되는 테스트라던지, 이런 부분은 과감히 통합 테스트를 적용 할 생각을 가지고 있어요ㅎㅎ

그리고 제가 해본 것들은 아주 작은 경험들이지만, 효율을 위해 `@Import` 라던지, `@Spy` 같은 것들을 적극 활용해서 Mocking을 피하는게 훨씬 효율적일 때가 있더라구요!

선영님은 Mock 테스트에 대해 평소 어떻게 생각 하시는지 궁금 하네요! 😄

shinsunyoung commented:
@MuseopKim 님! 제가 어제 잠들어버려서 ㅠ 이제 코멘트 드립니다! 
저는 외부 인프라스트럭처가 아닌 이상 Mock 사용은 자제하고 있어요! 

사실 저도 Mock을 많이 사용했었는데, Mock을 사용하게 되면 이미 개발자가 결과를 정해놓고 테스트를 하는게 되는데, 그게 의미있는 테스트일지에 대한 질문을 받은적이 있었어욥! 그 질문에 설득당해서 지금까지 Mock 사용을 의식적으로라도 자제해보는 연습을 하고 있습니당 ... 🙌

내가 테스트에서 중요하게 생각하는게 무엇인지, 상황이 어떤지에 따라 고민해보고 올바른 방법을 정하는게 좋을 것 같아요!

MuseopKim commented:
@shinsunyoung 아 그러셨군요!

주관이 들어가는 부분이라... 이건 여기저기서 계속 논란이 많은 걸로 알고 있어요ㅎㅎ 선영님께서 공유 해주신 내용에서처럼요!
읽어 보기 전에는 `Classicist vs Mockist` 제목만 보고 조금 극단적이라는 생각이 들었었는데... 읽어보니 내용은 아니더라구요ㅎㅎ
좋은 내용 공유 해주셔서 감사합니다! 😄

저는 이번에 루비콘에 들어오게 되면서 테크니컬 챌린지로 잡았던 몇가지가 있었는데. 그 중 하나가 TDD였어요. (실무에서 적용하긴 거의 불가능 할거 같고, 토이에서 시도해볼만 한 것 같아서...ㅎㅎ)

테스트를 먼저 짜고 그에 맞게 설계를 맞춰가보려고 하는데... 어려움이 많네요ㅎㅎ 시간도 생각보다 훨씬 오래 걸리구요 🤔
아마 완벽한 TDD는 (절대) 안될거 같고... 시도 해보고 반만 달성해도 좋은 경험이 될거라고 생각하고 있어요!

선영님께서 이미 많이 접해보신 내용이겠지만, 이번 프로젝트는 요렇게 개발 해보고 싶다는 의미가 잘 담겨 있어서 공유 해드립니다!
[TDD](https://ahea.wordpress.com/2018/09/10/%EC%84%A0%ED%83%9D%EC%9D%B4-%EC%95%84%EB%8B%8C-%ED%95%84%EC%88%98-tdd/) (내용은 좋은데 제목은... 좀 🤔)

MuseopKim committed:
Fix: DTO - `@NotNull`, `@Nullable` 적용

가독성을 위한 애노테이션 적용 및 테스트 코드를 위한 불필요한 기본 생성자 삭제

AccountRequest,
AccountSummary

[#8]

MuseopKim committed:
Fix: AccountSummary (DTO) - Role 필드 Enum으로 적용

[#5]

MuseopKim committed:
Fix: DTO로부터 Account를 생성하는 로직 변경

- 결합도를 낮추기 위해 엔티티에 DTO를 넘기는 것이 아니라, DTO에서 직접 생성하도록 변경
- submitCount, Role 기본 값을 빌더를 통해 통하도록 변경
- oauthId 필드명 oAuthId로 변경

[#5]

MuseopKim committed:
Fix: API 응답 객체 수정

- static method 기존 컨벤션대로 수정
- API 응답이라는 의미에 맞게 Result -> Response 네이밍 변경

[#5]

MuseopKim committed:
Fix: Role - 현재 사용하지 않는 값 제거 (ADMIN)

[#5]

MuseopKim committed:
Refactor: JwtTokenProvider - 메서드 리팩토링

- validateAuthorizationToken() : 반환형을 사용하지 않으므로 void 처리
- validateAuthorizationHeader() : 예외처리 로직 별도의 메서드로  분리

[#8]

MuseopKim committed:
Refactor: JwtTokenProvider - 토큰 prefix 상수로 분리

[#8]

MuseopKim committed:
Refactor: JwtTokenProvider - 헤더 검증 이후 다르다면 throw 를 바로 던지도록 수정

[#8]

MuseopKim committed:
Chore: JetBrains `@NotNull`, `@Nullable` 적용

[#8]

MuseopKim committed:
Merge pull request #9 from Lubycon/feature/issue-8

1차 리뷰 내용 반영 PR 요청 입니다!
